### PR TITLE
BUILD: add a windows meson CI job

### PR DESCRIPTION
**Pull request by mattip**
Using the scipy run as a template, add a windows meson CI run.

Replaces #23027

Does not use dev.py, that can be done in a subsequent PR.

### CODE REVIEW

**Comment by mattip** [comment]
Using the workflow from SciPy somehow gets meson into a mode where [it wants a gcc compiler](https://github.com/mattip/numpy/actions/runs/3957443510/jobs/6777820627#step:6:27). @rgommers any idea why running `meson setup build --prefix=$PWD\build` errors with
```
meson.build:28:4: ERROR: Problem encountered: NumPy requires GCC >= 8.4
```

**Comment by rgommers** [comment]
It finds:
```
C compiler for the host machine: gcc (gcc 8.1.0 "gcc.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0")
C linker for the host machine: gcc ld.bfd 2.30
C++ compiler for the host machine: c++ (gcc 8.1.0 "c++.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0")
C++ linker for the host machine: c++ ld.bfd 2.30
```

which is likely because since recently, MinGW is pre-installed on the GHA image. If you exactly copy the `install-rtools` step of the SciPy job, you should get gcc 10.3.0 instead (e.g., see [this recent CI log](https://github.com/scipy/scipy/actions/runs/3953254071/jobs/6769321998)).

**Comment by mattip** [comment]
I was using rtools in a previous iteration, but there were are many [build warnings](https://github.com/mattip/numpy/actions/runs/3957346042/jobs/6777602623#step:9:24) and an [error](https://github.com/mattip/numpy/actions/runs/3957346042/jobs/6777602623#step:9:7698). If this is required, I will have to go back and fix these problems. Is there a way to convince meson to use MSVC?

**Comment by rgommers** [comment]
Sure. At least, I had no problem using MSVC in https://github.com/scipy/scipy/pull/16957. Meson will just look for compilers and use the first/best one it finds. Unless we tell it explicitly through the `CC` and `CXX` env vars which ones to use.

**Comment by rgommers** [comment]
> The only compiler on the machine? 

I think the only one indeed. MinGW was only added to GHA after I touched that PR.

> Do I need to set just CC/CXX or CC_LD and CXX_LD as well?

I don't know what is best practice. Had a quick peek at `scikit-image`, and they do this scary-looking thing: https://github.com/scikit-image/scikit-image/blob/main/.github/workflows/wheel_tests_and_release.yml#L194-L212.

I have seen discussion in other places but cannot remember where. @eli-schwartz do you have a pointer on how best to select MSVC?

**Comment by eli-schwartz** [comment]
Meson will prefer MSVC as the first, best, most ideal compiler... if, and only if, you run meson from within a MSVC dev command prompt, which adds `cl.exe` to PATH as well as exporting a bunch of bizarre variables that makes it somehow work.

If that's not the case, then Meson tries to find a sensibly designed compiler, such as gcc or clang.

If none of those exist either, then Meson will search for vswhere, use it to detect a suitable vcvarsall batch script, run that batch script and print out the resulting environment variable state, then parse that into a dictionary of environment variables and update `os.environ` for the Meson process itself. At that point, it is (hopefully) the same as though you ran Meson from a MSVC dev command prompt, except that commands run *after* Meson (commands such as ninja!!!) still don't see the VS environment.

All the above is summarized, simply, as "Meson's `mesonlib.setup_vsenv()` internal utility routine".

setup_vsenv() is applied very early on before compilers are detected, and also saved as an attribute of the internal build data, so that `meson compile -C builddir` can check if a vsenv is needed, and also invoke setup_vsenv(). This unbreaks ninja, because ninja cannot compile anything with `cl.exe` unless the MSVC dev command prompt is set up. It would be nice if build.ninja accepted a setting to load an environment file containing the MSVC environment, but sadly it does not. There's an open ticket to add such a setting.

...

In general, setting the environment variable `$CC` is directly equivalent to setting `c = ...` in a machine file. Machine files take precedence. You can set either one to `cl.exe` and Meson won't try to search for gcc/clang, of course. It will just break if it cannot find it (for example because gcc was found by setup_vsenv).

...

You can force the vsenv to be set up by running Meson with the `--vsenv` option.

You can also run your entire CI workflow with the `ilammy/msvc-dev-cmd` github action.

**Comment by eli-schwartz** [comment]
> which is likely because since recently, MinGW is pre-installed on the GHA image.

GHA has, for absolutely ages, provided a Strawberry Perl. It's broken for numerous reasons, including the fact that it ships a private mingw toolchain -- a really outdated one that is also incompatible with the MSVC ABI and also incompatible with absolutely anything that modern mingw toolchains like msys2 try to provide.

Even if you don't find their Strawberry GCC, they ship a pkg-config.bat which is actually a perl program, and which badly reimplements pkg-config except broken. As a side effect, using pkg-config.bat will cause pkg-config dependencies to be resolved from inside the Strawberry Perl private libraries folder, even if you set up and used MSVC `cl.exe` -- I don't think I need to explain how broken that is.

Fortunately, Meson detects Strawberry pkg-config and emits a warning that it shall not be used -- instead, no pkg-config dependencies can be found until you provide a working pkg-config. ;)

...

tl;dr -- the situation is broken, Strawberry Perl should be deleted from GHA ideally and since GitHub doesn't want to do that everyone must work around it. Please, at all costs, avoid using Strawberry GCC.

**Comment by mattip** [comment]
Thanks! This is great information. Maybe it should appear in a FAQ soemwhere?

> You can also run your entire CI workflow with the ilammy/msvc-dev-cmd github action.

This seems like the easiest way forward here

**Comment by mattip** [comment]
It seems the cpp files are [getting a "/c" switch](https://github.com/mattip/numpy/actions/runs/3962630747/jobs/6789532159#step:8:304) which is leading to warnings and maybe some compilation problems?

**Comment by eli-schwartz** [comment]
I can't imagine why that would be a problem.
 That's the "compile without linking" option, which is central to incremental builds.

There's a `/Tc` option that tells the compiler to pretend that these files are (contrary to the file extension) actually C sources, and a `/Tp` option to do that for C++, but this isn't either one of those.

**Comment by rgommers** [comment]
Warnings are to be expected @mattip, best to ignore those and focus on the errors only. In this case, lots of this for openblas:
```
libopenblas.a(zsyrk_kernel_L.obj) : error LNK2001: unresolved external symbol ___chkstk_ms
```
That seems to be caused my mixing GCC and MSVC. This numpy build is picking up `msvc 19.29.30147`. And `libopenblas 0.3.18` was compiled with MinGW by Matthew. So that makes sense. 

The question is why the build picks up `libopenblas.a`, it should be picking up the shared library instead. I don't understand why `libopenblas.dll` is parked under `bin/` here rather than under `lib/`, that doesn't make much sense I think (from https://github.com/scipy/scipy-ci-artifacts/raw/main/openblas_32_if.zip):
image

**Comment by rgommers** [comment]
@mattip do we need the static library at all? I can't think of a reason to include it in that zipfile right now.

**Comment by mattip** [comment]
> Warnings are to be expected @mattip, best to ignore those

These warnings do not appear in the setup.py build, so it might be one source of problems.

> do we need the static library at all?

Ahh, maybe this package is not appropriate for the MSVC compiler: there should be a `*.lib` import library. I think the SciPy build uses mingw and not MSVC.

**Comment by rgommers** [comment]
The corresponding OpenBLAS builds from https://anaconda.org/multibuild-wheels-staging/openblas-libs/files do have an import library (`.lib`) and the same setup with the actual DLL in `bin/` rather than `lib/`. So I guess it's been like that for a long time, not sure why.

image


> I think the SciPy build uses mingw and not MSVC.

Good catch. The wheel builds do indeed use Mingw-w64. Building with MSVC + Intel Fortran is also supported, but this `openblas` tarball is just for numpy/scipy wheels so that's not very relevant.

**Comment by mattip** [comment]
Closing, I will reopen a better version of this
